sherpa/astro/io/tests/test_io_pha.py

```diff
@@ -0,0 +1,1366 @@
+    # (we are in a different directory to infile so we can't
+    # find these files).
+    #
+    assert outdata.response_ids == []
+    assert outdata.background_ids == []
+
+    assert outdata.get_arf() is None
+    assert outdata.get_rmf() is None
+    assert outdata.get_background() is None
+
+
+@requires_fits
+@requires_data
+def test_pha_write_xmm_grating(make_data_path, tmp_path):
+    """Check we can handle an XMM grating.
+
+    This has an AREASCAL column and WCS attached to the CHANNEL
+    colun. We don't guarantee the WCS info will be retained or
+    propagated.
+
+    """
+
+    def check_header(obj, roundtrip=False):
+        # check header for a selected set of keywords
+        hdr = obj.header
+
+        # selected OGIP keywords
+        assert "HDUNAME" not in hdr
+
+        assert hdr["HDUCLASS"] == "OGIP"
+        assert hdr["HDUCLAS1"] == "SPECTRUM"
+        assert hdr["HDUCLAS2"] == "NET"
+        assert hdr["HDUCLAS3"] == "COUNT"
+        if roundtrip:
+            assert hdr["HDUCLAS4"] == "TYPE:I"
+        else:
+            assert "HDUCLAS4" not in hdr
+        assert hdr["HDUVERS"] == "1.2.0"
+        assert "HDUVERS1" not in hdr
+
+        # a few header keywords to check they are handled,
+        # including data types (string, float, integer,
+        # logical).
+        #
+        assert hdr["OBJECT"] == "TW Hya"
+        assert hdr["TELESCOP"] == "XMM"
+        assert hdr["INSTRUME"] == "RGS1"
+        assert hdr["FILTER"] == "UNKNOWN"
+
+        assert not hdr["POISSERR"]
+        assert hdr["CHANTYPE"] == "PI"
+        assert hdr["DETCHANS"] == 3600
+
+        assert "EXPOSIRE" not in hdr
+
+        assert hdr["SYS_ERR"] == pytest.approx(0.0)
+        assert hdr["GROUPING"] == 0
+        assert "QUALITY" not in hdr
+        assert hdr["CORRSCAL"] == 1
+        assert "AREASCAL" not in hdr
+        assert "BACKSCAL" not in hdr
+
+        assert hdr["CORRFILE"] == "none"
+        for key in ["RESPFILE", "ANCRFILE", "BACKFILE"]:
+            assert key not in hdr
+
+        # WCS attached to the CHANNEL column
+        if backend_is("crates"):
+            assert "TCRYP1" not in hdr
+            assert "TCRVL1" not in hdr
+            assert "TLMIN1" not in hdr
+            assert "TLMAX1" not in hdr
+
+        elif backend_is("pyfits"):
+            assert hdr["TCTYP1"] == ""
+            assert hdr["TCUNI1"] == "Angstrom"
+            assert hdr["TCRPX1"] == 1
+            assert hdr["TCRVL1"] == pytest.approx(4.00500011)
+            assert hdr["TCDLT1"] == pytest.approx(0.01)
+
+            assert hdr["TLMIN1"] == 1
+            assert hdr["TLMAX1"] == 3600
+
+        else:
+            assert False  # programming error
+
+    def check_data(obj, roundtrip=False):
+        """Basic checks of the data"""
+
+        assert obj.staterror.min() == pytest.approx(1.0)
+        assert obj.staterror.sum() == pytest.approx(4496.9490242)
+        assert obj.staterror[1498] == pytest.approx(7.937253952)
```

Comment:
I don't ask to remove it - I do want to remark somewhere that tests tend to be super-extensive. If the min and the sum match, then this number will also match. Or maybe better test the min and [1498]. If those two are right, the sum will also be right. Nothing wrong with having four lines of test for the staterror column, but I feel one line would have been enough for our purposes and cover 99.99 % of the error that Sherpa could possibly introduce.

Reply:
I have seen too many strange errors.

Technically if the min and sum match we have not specified the whole dataset. I admit that adding a single value does not indicate we've truly caught all cases, and the delta improvement in our confidence that the code is working correctly is marginal at best, but the amount of time we've spent discussing this is more than the effort I spent on this.